### PR TITLE
Add SparsityTools::gather_sparsity_pattern()

### PR DESCRIPTION
**Pull request by davydden**
TODO:

- [x] try using `add_entries(row, ptr, ptr+num, true);`

### CODE REVIEW

**Comment by tjhei** [comment]
I don't understand the difference between this and the existing function in this namespace. Can you explain and put in the documentation?

**Comment by davydden** [comment]
> I don't understand the difference between this and the existing function in this namespace.

this one gathers sparsity, like `update_ghost_values()`, documentation is here, but perhaps it can be improved.

**Comment by tjhei** [comment]
> EDIT: existing one distributes sparsity (send to owners), this one adds to the local according to what other MPI processes have (ask owners to give data).

oh, that makes a lot more sense now.

File: include/deal.II/lac/sparsity_tools.h [hunk 1]
```diff
@@ -292,6 +292,28 @@ namespace SparsityTools
...
 
+  /**
+   * Gather rows in a dynamic sparsity pattern over MPI.
+   *
+   * @param dsp A dynamic sparsity pattern that has been built locally and which
+   * we need to extend according to the sparsity of rows stored on other MPI
+   * processes.
+   *
+   * @param rows_per_cpu A vector containing the number of of rows per CPU for
+   * determining ownership. This input should be the same on all MPI processes.
```
Comment (bangerth) [inline include/deal.II/lac/sparsity_tools.h]: Could you say how one would obtain this argument.

In fact, I'm not so happy with the way this argument is structured. It makes the implicit assumption that every process stores a contiguous range of rows, but we're tried to get away from this model and the way we initialize these objects these days is via `IndexSet` objects that may or may not be contiguous. I'd be much happier if the argument passed here were simply a vector of IndexSets. Assuming that these would correspond to locally owned DoFs, one can get this vector from DoFHandler.

File: source/lac/dynamic_sparsity_pattern.cc [hunk 2]
```diff
@@ -406,6 +406,26 @@ DynamicSparsityPattern::symmetrize()
...
+
+  if (rowset.size() > 0 && !rowset.is_element(row))
+    return;
+
+  const size_type rowindex =
+    rowset.size() == 0 ? row : rowset.index_within_set(row);
+
+  AssertIndexRange(rowindex, lines.size());
+  std::vector<size_type> entries;
+  lines[rowindex].entries.swap(entries);
```
Comment (bangerth) [inline source/lac/dynamic_sparsity_pattern.cc]: better:
```
  lines[rowindex].entries = std::vector<size_type>();
```
This uses the move assignment operator and has the same effect.

File: include/deal.II/lac/sparsity_tools.h (hunk 1 above)
Comment (davydden) [reply include/deal.II/lac/sparsity_tools.h]: > I'd be much happier if the argument passed here were simply a vector of IndexSets

I can take a `const IndexSet & owned_rows` and deduce what I need form there. I will still support only contiguous ranges at this point.

Comment (bangerth) [reply include/deal.II/lac/sparsity_tools.h]: That's ok. Someone could then later implement the non-contiguous case, but at least we wouldn't have to change the interface. In the meantime, you can just check `index_set.is_contiguous()`.

Vector vs single index set: That sounds reasonable. But doesn't the current object knows its local index set already?

Comment (tjhei) [reply include/deal.II/lac/sparsity_tools.h]: > I can take a `const IndexSet & owned_rows` and deduce what I need form there. I will still support only contiguous ranges at this point. Would you be ok with this solution?

Or a std::vector<IndexSet> as returned by ``locally_owned_dofs_per_processor()``

Comment (davydden) [reply include/deal.II/lac/sparsity_tools.h]: > a std::vector as returned by locally_owned_dofs_per_processor()

i intend to use it outside of DoFHandler context, so I would rather not take an `std::vector<IndexSet>` here.

Comment (davydden) [reply include/deal.II/lac/sparsity_tools.h]: Nevermind, i will update and add an assert.

**Comment by davydden** [comment]
@tjhei @bangerth I updated the interface.

**Pull request merged**